### PR TITLE
CLOUD-544 Use version service

### PR DESCRIPTION
**Pull request by blez**
[CLOUD-544](https://jira.percona.com/browse/CLOUD-544)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| recreate | failed |
| scaling | failed |
| big-data | failed |
| storage | failed |
| scaling-proxysql | failed |
| restore-to-encrypted-cluster | failed |
| limits | failed |
| upgrade | failed |
| self-healing | failed |
| monitoring | failed |
| upgrade-consistency | failed |
| self-healing-advanced | failed |
| demand-backup | failed |
| security-context | failed |
| monitoring-2-0 | failed |
| affinity | failed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |
| one-pod | failed |
| auto-tuning | failed |
| operator-self-healing | failed |
| proxysql-sidecar-res-limits | failed |

File: pkg/controller/pxc/version.go [hunk 1]
```diff
@@ -71,80 +79,109 @@ func (r *ReconcilePerconaXtraDBCluster) sheduleEnsurePXCVersion(cr *api.PerconaX
...
+		},
+	); err != nil {
+		return fmt.Errorf("get pod list: %v", err)
 	}
 
-	return vr
-}
+	user := "root"
+	for _, pod := range list.Items {
+		dbatabase, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, pod.Status.PodIP, 3306)
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/pxc/version.go:161]: https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/f835406b08d9fab4f81619c8291666e81e309ede

File: pkg/controller/pxc/controller.go [hunk 2]
```diff
@@ -133,6 +136,7 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 		RequeueAfter: time.Second * 5,
 	}
 	// Fetch the PerconaXtraDBCluster instance
+	r.statusMutex.Lock()
 	o := &api.PerconaXtraDBCluster{}
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/pxc/controller.go:142]: Great spotted!
https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/c75d30748815b4722f3aa8b7843a8a7f2307c4b9

File: pkg/controller/pxc/version.go [hunk 3]
```diff
@@ -71,80 +79,109 @@ func (r *ReconcilePerconaXtraDBCluster) sheduleEnsurePXCVersion(cr *api.PerconaX
...
 }
 
-func (vs VersionServiceMock) CheckNew() VersionResponse {
-	vr := VersionResponse{
-		PXCImage:      "percona/percona-xtradb-cluster-operator:1.4.0-pxc8.0",
-		PXCVersion:    "8.0.18-9.3",
-		BackupImage:   "perconalab/percona-xtradb-cluster-operator:master-pxc8.0",
-		PMMImage:      "perconalab/percona-xtradb-cluster-operator:master-pmm",
-		ProxySqlImage: "perconalab/percona-xtradb-cluster-operator:master-proxysql",
+func (r *ReconcilePerconaXtraDBCluster) updateVersion(cr *api.PerconaXtraDBCluster, sfs api.StatefulApp) error {
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/pxc/version.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/8f56ef4a9a606fab38fc19adadae0f8cc4d63c79

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

File: pkg/controller/pxc/version.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (qJkee) [reply pkg/controller/pxc/version.go:161]: still there =)

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | failed |
| big-data | failed |
| init-deploy | failed |
| storage | failed |
| recreate | failed |
| limits | failed |
| restore-to-encrypted-cluster | failed |
| self-healing | failed |
| scaling-proxysql | failed |
| self-healing-advanced | failed |
| upgrade | passed |
| demand-backup | failed |
| monitoring | failed |
| operator-self-healing | failed |
| scheduled-backup | failed |
| monitoring-2-0 | failed |
| demand-backup-encrypted-with-tls | failed |
| affinity | failed |
| one-pod | failed |
| upgrade-consistency | passed |
| security-context | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| scaling | failed |
| storage | failed |
| recreate | failed |
| big-data | failed |
| self-healing | failed |
| limits | failed |
| scaling-proxysql | failed |
| self-healing-advanced | failed |
| restore-to-encrypted-cluster | failed |
| monitoring | failed |
| operator-self-healing | failed |
| demand-backup | failed |
| upgrade | failed |
| monitoring-2-0 | failed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |
| affinity | passed |
| one-pod | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| recreate | failed |
| big-data | failed |
| storage | failed |
| scaling | failed |
| self-healing | failed |
| scaling-proxysql | failed |
| restore-to-encrypted-cluster | failed |
| upgrade | passed |
| self-healing-advanced | failed |
| demand-backup | failed |
| limits | passed |
| operator-self-healing | failed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| security-context | failed |
| monitoring-2-0 | failed |
| affinity | passed |
| one-pod | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | failed |
| storage | passed |
| init-deploy | failed |
| scaling | failed |
| recreate | passed |
| self-healing | failed |
| scaling-proxysql | passed |
| limits | passed |
| self-healing-advanced | failed |
| upgrade | failed |
| operator-self-healing | failed |
| monitoring | passed |
| upgrade-consistency | passed |
| restore-to-encrypted-cluster | failed |
| security-context | failed |
| demand-backup | failed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | failed |
| scheduled-backup | failed |
| auto-tuning | passed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| big-data | failed |
| init-deploy | failed |
| recreate | passed |
| scaling | failed |
| limits | passed |
| self-healing | failed |
| monitoring | passed |
| scaling-proxysql | failed |
| restore-to-encrypted-cluster | failed |
| upgrade | passed |
| self-healing-advanced | passed |
| demand-backup | failed |
| monitoring-2-0 | passed |
| operator-self-healing | failed |
| upgrade-consistency | passed |
| security-context | failed |
| affinity | passed |
| one-pod | failed |
| scheduled-backup | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | failed |
| init-deploy | failed |
| recreate | failed |
| big-data | failed |
| storage | passed |
| limits | failed |
| scaling-proxysql | passed |
| self-healing | failed |
| monitoring | failed |
| upgrade | failed |
| restore-to-encrypted-cluster | failed |
| self-healing-advanced | passed |
| demand-backup | failed |
| upgrade-consistency | passed |
| monitoring-2-0 | failed |
| security-context | failed |
| scheduled-backup | failed |
| operator-self-healing | passed |
| affinity | passed |
| one-pod | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| big-data | failed |
| scaling | failed |
| storage | passed |
| recreate | passed |
| limits | passed |
| scaling-proxysql | failed |
| monitoring | passed |
| self-healing | failed |
| monitoring-2-0 | passed |
| restore-to-encrypted-cluster | failed |
| upgrade | failed |
| self-healing-advanced | passed |
| demand-backup | failed |
| affinity | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| security-context | failed |
| one-pod | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| scaling | passed |
| recreate | passed |
| big-data | failed |
| storage | failed |
| self-healing | failed |
| scaling-proxysql | passed |
| limits | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | failed |
| self-healing-advanced | passed |
| upgrade-consistency | passed |
| monitoring | passed |
| security-context | failed |
| demand-backup | failed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| affinity | passed |
| one-pod | failed |
| auto-tuning | failed |
| scheduled-backup | failed |
| proxysql-sidecar-res-limits | failed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| big-data | failed |
| limits | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| recreate | passed |
| scaling | passed |
| storage | passed |
| big-data | failed |
| scaling-proxysql | passed |
| self-healing | passed |
| limits | passed |
| monitoring | passed |
| upgrade | failed |
| restore-to-encrypted-cluster | failed |
| self-healing-advanced | passed |
| monitoring-2-0 | failed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| demand-backup | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| scheduled-backup | passed |
| demand-backup-encrypted-with-tls | failed |
| upgrade-consistency | failed |

File: pkg/controller/pxc/vs.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+func getVersion(versions map[string]Version) (string, error) {
+	if len(versions) == 0 {
+		return "", fmt.Errorf("failed to get version from empty map")
+	}
+
+	for k := range versions {
+		return k, nil
+	}
+	return "", nil
+}
```
Comment (dAdAbird) [inline pkg/controller/pxc/vs.go:78]: 1. Why do we return a random version?
2. Shouldn't be "status" and "critical" fields taken into account?

Comment (delgod) [reply pkg/controller/pxc/vs.go:78]: only the one version must be returned, we should choose the first in the list

File: pkg/controller/pxc/vs.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	BackupImage     string `json:"backupImage,omitempty"`
+	BackupVersion   string `json:"backupVersion,omitempty"`
+	ProxySqlImage   string `json:"proxySqlImage,omitempty"`
+	ProxySqlVersion string `json:"proxySqlVersion,omitempty"`
+	PMMImage        string `json:"pmmImage,omitempty"`
+	PMMVersion      string `json:"pmmVersion,omitempty"`
+}
+
+type VersionService interface {
+	Apply(desiredVersion string, currentVersion string) (DepVersion, error)
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/pxc/vs.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/fd49fa46b5df0f053591c2c0cc3fdebd6a0f1c91

File: pkg/controller/pxc/vs.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
+package pxc
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+func (vs VersionServiceMock) Apply(desiredVersion string, currentVersion string) (DepVersion, error) {
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/pxc/vs.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/9a5171fcfeb8261a6d65e52e39349f7e43c09650

File: pkg/controller/pxc/controller.go [hunk 7]
```diff
@@ -140,6 +149,10 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 		RequeueAfter: time.Second * 5,
 	}
 	// Fetch the PerconaXtraDBCluster instance
+	r.statusMutex.Lock()
+	defer r.statusMutex.Unlock()
+	defer atomic.StoreInt32(&r.updateSync, updateDone)
```
[the remark this answers is not in the document]
Comment (blez) [reply pkg/controller/pxc/controller.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/b19ac2d6790ee3c60c1a48e71534a8d774313e20

File: pkg/controller/pxc/vs.go (hunk 4 above)
Comment (blez) [reply pkg/controller/pxc/vs.go:78]: Exactly, we assume that in response we have just one version

Comment (blez) [reply pkg/controller/pxc/vs.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/446/commits/80c40ef050a99f663325d529f15826d3694b5f64

File: pkg/controller/pxc/controller.go [hunk 8]
```diff
@@ -140,6 +149,13 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 		RequeueAfter: time.Second * 5,
 	}
 	// Fetch the PerconaXtraDBCluster instance
+	r.statusMutex.Lock()
+	defer r.statusMutex.Unlock()
+	// this is for ensure that reconcile loop will always run
+	// after version service cron job and prevent situation when a few cron
+	// jobs will be run one by one
```
Comment (dAdAbird) [inline pkg/controller/pxc/controller.go:156]:
```suggestion
	// we have to be sure the reconcile loop will be run at least once
	// in-between any version service jobs (hence any two vs jobs shouldn't be run sequentially). 
	// the version service job sets the state to  `updateWait` and the next job can be run only
	// after the state was dropped to`updateDone` again
```

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | failed |
| recreate | failed |
| big-data | failed |
| init-deploy | failed |
| storage | failed |
| scaling-proxysql | failed |
| limits | passed |
| upgrade | failed |
| restore-to-encrypted-cluster | failed |
| monitoring | failed |
| demand-backup | failed |
| self-healing | failed |
| monitoring-2-0 | failed |
| scheduled-backup | failed |
| self-healing-advanced | failed |
| affinity | passed |
| operator-self-healing | failed |
| demand-backup-encrypted-with-tls | failed |
| one-pod | passed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| recreate | failed |
| storage | failed |
| big-data | failed |
| scaling | failed |
| restore-to-encrypted-cluster | failed |
| self-healing | failed |
| limits | passed |
| demand-backup | failed |
| scaling-proxysql | failed |
| self-healing-advanced | failed |
| operator-self-healing | failed |
| upgrade | failed |
| monitoring | failed |
| monitoring-2-0 | failed |
| scheduled-backup | failed |
| affinity | passed |
| demand-backup-encrypted-with-tls | failed |
| one-pod | passed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |
| upgrade-consistency | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |
| recreate | passed |
| init-deploy | passed |
| scaling | failed |
| big-data | passed |
| self-healing | passed |
| limits | passed |
| restore-to-encrypted-cluster | passed |
| scaling-proxysql | passed |
| monitoring | passed |
| upgrade | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | failed |
| upgrade-consistency | passed |
| security-context | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | passed |
| demand-backup-encrypted-with-tls | passed |
| proxysql-sidecar-res-limits | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| scaling | failed |
| recreate | passed |
| storage | passed |
| big-data | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| limits | passed |
| monitoring | passed |
| restore-to-encrypted-cluster | passed |
| upgrade | passed |
| self-healing-advanced | passed |
| upgrade-consistency | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | failed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | failed |
| demand-backup-encrypted-with-tls | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| scaling | passed |
| big-data | passed |
| storage | passed |
| recreate | failed |
| limits | failed |
| self-healing | failed |
| restore-to-encrypted-cluster | passed |
| scaling-proxysql | passed |
| self-healing-advanced | failed |
| operator-self-healing | passed |
| upgrade | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| demand-backup | failed |
| scheduled-backup | failed |
| affinity | passed |
| one-pod | failed |
| demand-backup-encrypted-with-tls | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |

**Pull request merged**